### PR TITLE
Segmentation hfs

### PR DESCRIPTION
**Pull request by Yujun-Shi**
### This pullrequest changes
Adding a very efficient module for image segmentation.
related #1401 #1410

### CODE REVIEW

**Comment by mshabunin** [comment]
@Yujun-Shi , it is not necessary to create new pull requests each time you want to update the code. Just push new commits to the same branch (_segmentation_hfs_) and pull request will get updated automatically.

**Comment by Yujun-Shi** [comment]
@mshabunin yes, I know. The previous two pull requests can't be rerun because their branch name is "master". The branch name of this pull request is not "master" any more, so I won't create another pull request again.  Thanks you for your attention!

File: modules/hfs/include/opencv2/hfs.hpp [hunk 1]
```diff
@@ -0,0 +1,200 @@
+/*M///////////////////////////////////////////////////////////////////////////////////////
```
Comment (sovrasov) [inline modules/hfs/include/opencv2/hfs.hpp]: Please use short license header for the new files (except samples):
```
// This file is part of OpenCV project.
// It is subject to the license terms in the LICENSE file found in the top-level directory
// of this distribution and at http://opencv.org/license.html.
/*
```

File: modules/hfs/include/opencv2/hfs.hpp [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+// loss of use, data, or profits; or business interruption) however caused
+// and on any theory of liability, whether in contract, strict liability,
+// or tort(including negligence or otherwise) arising in any way out of
+// the use of this software, even if advised of the possibility of such damage.
+//
+//M*/
+
+#ifndef _OPENCV_HFS_SEGMENT_HPP_
+#define _OPENCV_HFS_SEGMENT_HPP_
+#ifdef __cplusplus
```
Comment (sovrasov) [inline modules/hfs/include/opencv2/hfs.hpp]: We don't need such check anymore. It can be removed.

File: modules/hfs/include/opencv2/hfs.hpp [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+/** @brief do segmentation
+* @param src: the input image
+* @param ifDraw: if draw the image in the returned Mat. if this parameter is false,
+* then the content of the returned Mat is a matrix of index, describing the region
+* each pixel belongs to. And it's data type is CV_16U. If this parameter is true,
+* then the returned Mat is a segmented picture, and color of each region is the
+* average color of all pixels in that region. And it's data type is the same as
+* the input image
+*/
+virtual Mat performSegment(const Mat& src, bool ifDraw = true) = 0;
```
Comment (sovrasov) [inline modules/hfs/include/opencv2/hfs.hpp]: `InputArray` and `OutputArray` should be used in public interfaces.

File: modules/hfs/samples/example.cpp [hunk 4]
```diff
@@ -0,0 +1,50 @@
+#include "opencv2/hfs.hpp"
+#include "opencv2/opencv.hpp"
+
+using namespace cv;
+using namespace cv::hfs;
+int main()
+{
+    String str = "data/000";
```
Comment (sovrasov) [inline modules/hfs/samples/example.cpp]: This path will be invalid in most cases. Use command line arguments to provide the path to an input image.

File: modules/hfs/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,6 @@
+if(NOT HAVE_CUDA)
+  ocv_module_disable(hfs)
```
Comment (sovrasov) [inline modules/hfs/CMakeLists.txt]: Does this module strongly depends on CUDA?

Comment (Yujun-Shi) [reply modules/hfs/CMakeLists.txt]: yes, it can't run without cuda

File: modules/hfs/samples/example.cpp (hunk 4 above)
Comment (Yujun-Shi) [reply modules/hfs/samples/example.cpp]: alright, I see :)

**Comment by Yujun-Shi** [comment]
@sovrasov I've already modified my code as you demanded. Could you please review it again. Thanks a lot!

File: modules/hfs/samples/example.cpp [hunk 6]
```diff
@@ -0,0 +1,88 @@
+/*M///////////////////////////////////////////////////////////////////////////////////////
```
Comment (sovrasov) [inline modules/hfs/samples/example.cpp]: In samples the license header can be removed at all.

Comment (Yujun-Shi) [reply modules/hfs/samples/example.cpp]: Oh sorry, I seem to misinterpret your comments :)

**Comment by Yujun-Shi** [comment]
@sovrasov I'm sorry, I removed license comment in the sample file, and pushed my commit, but all of the buildbot cases generate error like this:

try to acquire commands lock
commands lock was acquired in 0.000582218170166 sec

git pull -v "https://github.com/Yujun-Shi/opencv_contrib.git" "segmentation_hfs"
 in dir /build/precommit-contrib_cuda_linux64/opencv_contrib (timeout 1200 secs)
 watching logfiles {}
 argv: git pull -v "https://github.com/Yujun-Shi/opencv_contrib.git" "segmentation_hfs"
 environment:
  HOME=/home/build
  OPENCV_GIT_URL=git://code.ocv/opencv
  PATH=/env/bin:/app/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games
  PWD=/build/precommit-contrib_cuda_linux64/opencv_contrib
  SHELL=/bin/bash
  TERM=xterm
  USER=build
 using PTY: False
POST git-upload-pack (947 bytes)
POST git-upload-pack (652 bytes)
From https://github.com/Yujun-Shi/opencv_contrib
 * branch            segmentation_hfs -> FETCH_HEAD
Auto-merging modules/README.md
CONFLICT (content): Merge conflict in modules/README.md
Automatic merge failed; fix conflicts and then commit the result.
program finished with exit code 1
elapsedTime=2.502987


commands lock was released

Do you have any idea what just happened?
thanks a lot!!!

File: modules/README.md [hunk 7]
```diff
@@ -9,63 +9,34 @@ $ cmake -D OPENCV_EXTRA_MODULES_PATH=<opencv_contrib>/modules -D BUILD_opencv_<r
 ```
 
 - **aruco**: ArUco and ChArUco Markers -- Augmented reality ArUco marker and "ChARUco" markers where ArUco markers embedded inside the white areas of the checker board.
-
 - **bgsegm**: Background segmentation algorithm combining statistical background image estimation and per-pixel Bayesian segmentation.
-
```
Comment (alalek) [inline modules/README.md]: You need to:
- remove unrelated changes
- squash your commits
- rebase on upstream "master" branch (last two steps can be don via "git rebase -i upstream/master", where upstream is configured remote for opencv/opencv_contrib (see GitHub help))

Comment (Yujun-Shi) [reply modules/README.md]: Thanks a lot! Now everything works out!!! 👍

**Comment by alalek** [comment]
BTW, CUDA-only module perhaps will not be accepted. Reference C++ implementation is required, at least for safe refactoring / adding tests.

**Comment by Yujun-Shi** [comment]
@alalek But according to @sovrasov , there will be another module that contains all CUDA stuff, and this PR can be merged into that, right?

**Comment by alalek** [comment]
He mentions special [CUDA modules](https://github.com/opencv/opencv/tree/3.3.1/modules) with CUDA-implementation (`cv::cuda::addWeighted`) of existed OpenCV functions (`cv::addWeighted`).
Tests in these modules checks results similarity between these implementations (CUDA and C++).
Here we have new algorithm implemented on CUDA but without reference C++ code.

You need to rebase your commits on current master (rebase and squash into one). Also it is better to remove unrelated changes (see `modules/README.md) file to avoid future conflicts).

**Comment by Yujun-Shi** [comment]
@alalek OK, I'll try to add the CPU implementation as soon as possible and add it to the module in the next commit. Thank you for your help!

**Comment by Yujun-Shi** [comment]
@alalek hi, sorry for the delay. Now the CPU interface is added. Could you please check it out again.

File: modules/hfs/src/or_utils/or_vector.hpp [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+    __CV_CUDA_HOST_DEVICE__ Vector2(const Vector2_<T> &v)
+    {
+        this->x = v.x;
+        this->y = v.y;
+    }
+
+    __CV_CUDA_HOST_DEVICE__ friend Vector2<T> &operator /= (Vector2<T> &lhs, T d)
+    {
+        if (d == 0) return lhs; lhs.x /= d; lhs.y /= d; return lhs;
```
Comment (alalek) [inline modules/hfs/src/or_utils/or_vector.hpp]: Static code analyzers / modern compilers dislike this line (they warn about missing `{}`).
Please split this line on 2-3 lines.

```
modules/hfs/src/or_utils/or_vector.hpp: In function 'cv::hfs::orutils::Vector2<T>& cv::hfs::orutils::operator/=(cv::hfs::orutils::Vector2<T>&, T)':
modules/hfs/src/or_utils/or_vector.hpp:46:9: warning: this 'if' clause does not guard... [-Wmisleading-indentation]
         if (d == 0) return lhs; lhs.x /= d; lhs.y /= d; return lhs;
         ^~
```

Comment (Yujun-Shi) [reply modules/hfs/src/or_utils/or_vector.hpp]: ok, I'll fix it right away :)

**Comment by Yujun-Shi** [comment]
@alalek Alright, I just fixed that static check warning.

**Comment by alalek** [comment]
GPU and CPU results are different (sample with 000.jpg input):
segment_default_gpu
segment_default_cpu

```
```

**Comment by Yujun-Shi** [comment]
@alalek Well, In the first picture, the GPU version obviously failed. I've run the module several times in different configuration, and it worked out just fine. I'll try to figure out what is going on here. Also, could you pls tell me more about your configurations? Something like which OS did you use here, which compiler did you use. Thanks a lot!

**Comment by alalek** [comment]
@Yujun-Shi I updated my comment above.

File: modules/hfs/include/opencv2/hfs.hpp [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+For further details about the algorithm, please refer to the original paper:
+Hierarchical Feature Selection for Efficient Image Segmentation, ECCV 2016
+
+*/
+
+class CV_EXPORTS_W HfsSegment : public Algorithm {
+public:
+
+//! @addtogroup hfs
+//! @{
```
Comment (alalek) [inline modules/hfs/include/opencv2/hfs.hpp]: "addtogroup" should be outside of class.

File: modules/hfs/src/cuda/gslic_seg_engine_gpu.cu [hunk 10]
```diff
@@ -0,0 +1,326 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+#ifdef _HFS_CUDA_ON_
```
Comment (alalek) [inline modules/hfs/src/cuda/gslic_seg_engine_gpu.cu]: I believe this guard is not necessary in ".cu" files.

File: modules/hfs/src/magnitude/magnitude.cpp [hunk 11]
```diff
@@ -0,0 +1,216 @@
+#include "magnitude.hpp"
```
Comment (alalek) [inline modules/hfs/src/magnitude/magnitude.cpp]: Add license header and
```
#include "../precomp.hpp"
```

File: modules/hfs/src/magnitude/magnitude.hpp [hunk 12]
```diff
@@ -0,0 +1,47 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+#ifndef _OPENCV_MAGNITUDE_HPP_
+#define _OPENCV_MAGNITUDE_HPP_
+#include "../precomp.hpp"
```
Comment (alalek) [inline modules/hfs/src/magnitude/magnitude.hpp]: Don't use "precomp.hpp" from `.hpp` files

File: modules/hfs/src/slic/gslic_engine.cpp [hunk 13]
```diff
@@ -0,0 +1,63 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+#ifdef _HFS_CUDA_ON_
+
+#include "../precomp.hpp"
```
Comment (alalek) [inline modules/hfs/src/slic/gslic_engine.cpp]: precomp.hpp should be before any "#if".

File: modules/hfs/src/slic/slic.cpp [hunk 14]
```diff
@@ -0,0 +1,219 @@
+#include "slic.hpp"
+#include "../precomp.hpp"
```
Comment (alalek) [inline modules/hfs/src/slic/slic.cpp]: precomp.hpp must be first.

**Comment by Yujun-Shi** [comment]
@alalek sorry for the delay. Now everything except that CUDA issue is fixed. Could you pls check it again. Thx!
